%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Snakefiles as Python-Code}
{   
	\usebackgroundtemplate{
		\vbox to \paperheight{\vfil\hbox to \paperwidth{\hfil\includegraphics[height=\paperheight]{logos/Text-x-python.svg.png}\hfil}\vfil}
		% source: https://en.m.wikipedia.org/wiki/File:Text-x-python.svg
	}
	\frame{
		\frametitle{Python Code in your Snakefile}
		\begin{mdframed}[tikzsetting={draw=white,fill=white,fill opacity=0.8,
				line width=0pt},backgroundcolor=none,leftmargin=0,
			rightmargin=150,innertopmargin=4pt,roundcorner=10pt]
			\tableofcontents[currentsection,sections={1-4},hideothersubsections]
		\end{mdframed}
	}
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
  \frametitle{What is this about?}
   \begin{question}[Questions]
   	 \begin{itemize}
        \item How can I automatically manage dependencies and outputs?
        \item How can I use Python code to add features to my workflow?
     \end{itemize}
   \end{question}
   \begin{docs}[Objectives]
   	  \begin{enumerate}
                      \item Use variables, functions, and imports in a Snakefile. 
                      \item Learn to use the \texttt{run} action to execute Python code as an action.
       \end{enumerate}
   \end{docs}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Python Code in Snakefiles}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
  \frametitle{Why Python in a Snakefile?}
  Sometimes we do \emph{not} want to run $3^{\mathsf{rd}}$ party code, but run the occasional script for data manipulation   or plotting or \ldots 
  \pause
  \begin{docs}
  	\altverb{Snakefile}s are Python code (albeit special Python code)!)
  \end{docs}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Supported Functions}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{Using functions in Snakefiles}
  We already met the \altverb{expand()}-function. Let's revisit it in detail!
  \begin{task}
  	Start Python and follow along!
  \end{task}
  \begin{lstlisting}[language=Python,style=Python]
from snakemake.io import expand, glob_wildcards
  \end{lstlisting}
   \altverb{expand()} is used frequently, to expand a \Snakemake{} wildcard(s) into a set of filenames:
  \begin{lstlisting}[language=Python,style=Python]
>>> expand('folder/{wildcard1}_{wildcard2}.txt',
...        wildcard1=['a', 'b', 'c'],
...        wildcard2=[1, 2, 3])
  \end{lstlisting}
\end{frame}

% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{Using functions in \altverb{Snakefile}s II}
  \begin{question}
  	What is the result?
  \end{question}
  \pause
  \begin{hint}[Answer:]
  	Every permutation of \altverb{wildcard1} and \altverb{wildcard2}!
  \end{hint}
  \pause
  Let us try \altverb{glob_wildcards()}, now:
  \begin{lstlisting}[language=Python,style=Python]
glob_wildcards('data/samples/{replicas}.fastq').replicas
  \end{lstlisting}
  \pause
  Wow! So, easy!
  \begin{question}
  	What happened? What happens if you leave \altverb{.replicas} away?
  \end{question}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\texttt{expand()} vs. \texttt{glob\_wildcards()}}
  \begin{exampleblock}{The Difference}
    \begin{columns}
      \begin{column}{0.45\textwidth}
        \begin{itemize}
         \item \altverb{expand()} is usefull to yield all permutations of wildcards
         \item \altverb{expand()} is oblivious to the underlying file system
        \end{itemize}

      \end{column}
      \begin{column}{0.45\textwidth}
        \begin{itemize}
         \item \altverb{glob_wildcards} infers files from wildcards
         \item \altverb{glob_wildcards} must be used with care - users tend to name files arbitrarily
        \end{itemize}

      \end{column}
    \end{columns}
  \end{exampleblock}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Python Code as Actions}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\Interlude{Meeting the \texttt{run} Action}}
  \vspace{-0.5em}
  It is possible to run Python snippets in \Snakemake{}. For this, we have to replace the \altverb{shell:} action by a \altverb{run:} action:\vspace{-0.5em}
  \begin{task}
  	Imaging this code in a \altverb{Snakefile}:
  \end{task}\vspace{-0.5em}
  \begin{lstlisting}[language=Python,style=Python, basicstyle=\footnotesize]
# at the top of the file
import glob

# add this wherever
rule print_sample_names:
    run:
        print('These are all the samples names:')
        for sample in glob.glob('data/samples/*.fastq'):
            print(sample)

  \end{lstlisting}\vspace{-0.5em}
  \pause\footnotesize
  \begin{hint}[The idea:]
     This allows for Python snippets in rules without saving Python files, e.\,g. for short data mangling or plotting.
  \end{hint}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Snakemake and external Scripts}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\Snakemake{} Specials for Python}
  Remember this part:
  \begin{lstlisting}[language=Python,style=Python]
shell:  'some_command -i {input} -o {output}'
  \end{lstlisting}
  \begin{question}
  	Isn't this rather tedious, considering that \Snakemake{} is written in Python? Could we forward rule specific information to our scripts?
  \end{question}
\end{frame}

\begingroup
\setbeamertemplate{footline}{}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\Snakemake's features for external Scripts}
  In addition to the \altverb{shell} and \altverb{run} directives, \Snakemake{} offers a \altverb{script} directive, too:
  \begin{lstlisting}[language=Python,style=Python]
rule NAME:
    input:
        "path/to/inputfile",
    output:
        "path/to/outputfile",
    script:
        "scripts/script.py"
  \end{lstlisting}
  This lets you use (in \altverb{script.py}):
  \begin{lstlisting}[language=Python,style=Python]
def do_something(data_path, out_path):
    # python code

do_something(snakemake.input[0], snakemake.output[0])
  \end{lstlisting}
\end{frame}
\endgroup

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\Interlude{The same feature is offered for R}}
  \begin{lstlisting}[language=R,style=R]
do_something <- function(data_path, output_path) {
    # R code
}

do_something(snakemake@input[[1]], snakemake@output[[1]])
  \end{lstlisting}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\Interlude{Debugging external Scripts}}
  \begin{exampleblock}{Debugging Python Scripts}
  To debug Python scripts, you can invoke \Snakemake{} with
  \begin{lstlisting}[language=Bash, style=Shell]
$ snakemake --debug
  \end{lstlisting}
  As always in Python, you may use \altverb{print()} and \altverb{logging()} functions which may write to a log file during the run to check whether the script behaves as expected.
  \end{exampleblock}
  \pause
  \begin{exampleblock}{Debugging R Scripts}
  To debug R scripts, you can save the workspace with \altverb{save.image()}, and invoke R after \Snakemake{} has terminated. 
  \end{exampleblock}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{We want to plot our Results!}
  We now add the following rule to your \altverb{Snakefile} we could plot our variant statistics (see \altverb{06_Snakefile} in our tutorial folder):
  \begin{lstlisting}[language=Python,style=Python]
rule plot_quals:
    input:
        "calls/all.vcf"
    output:
        "plots/quals.svg"
    @script:@
        "scripts/plot-quals.py"
  \end{lstlisting}
  It uses the \altverb{script} directive and expects a directory \altverb{scripts} relative to our \altverb{Snakefile}.
  \pause
  \begin{task}
  	Create the missing directory "\altverb{scripts}".
  \end{task}
\end{frame}

\setcounter{preframe_handson}{\value{handson}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\HandsOn{Adding the Plotting Script}}
  \setcounter{handson}{\value{preframe_handson}}
  \begin{task}
  	Open an editor and follow along.
  \end{task}
  \begin{onlyenv}<1| handout:0>
   First, we start with importing a plotting library; save the script under \altverb{scripts/plot-quals.py}:
   \begin{lstlisting}[language=Python,style=Python]
@import matplotlib@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<2| handout:0>
   We enforce writing to suppress the interactive mode:
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
@matplotlib.use("Agg")@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<3| handout:0>
   We want an alternative interface:
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
@import matplotlib.pyplot as plt@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<4| handout:0>
   \ldots and need to import a function to read our variant file:
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
@from pysam import VariantFile@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<5| handout:0>
   Let's start a list comprehension \ldots
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from pysam import VariantFile

@quals = [@
   \end{lstlisting}
  \end{onlyenv}
   \begin{onlyenv}<6| handout:0>
   \ldots to construct a list of quality scores from our previous outputs:
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from pysam import VariantFile

@quals = [record.qual for record 
         in VariantFile(snakemake.input[0])]@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<7| handout:0>
   Finally, plot a histogram \ldots
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from pysam import VariantFile

quals = [record.qual for record 
         in VariantFile(snakemake.input[0])]
         
@plt.hist(quals)@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<8| handout:0>
   \ldots and save the figure.
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from pysam import VariantFile

quals = [record.qual for record 
         in VariantFile(snakemake.input[0])]
         
plt.hist(quals)
@plt.savefig(snakemake.output[0])@
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<9| handout:1>
   This should be our final script:
   \begin{lstlisting}[language=Python,style=Python]
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from pysam import VariantFile

quals = [record.qual for record 
         in VariantFile(snakemake.input[0])]
         
plt.hist(quals)
plt.savefig(snakemake.output[0])
   \end{lstlisting}
  To be saved as \altverb{scripts/plot-quals.py}.
  \end{onlyenv}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\HandsOn{One more thing \ldots}}
  \begin{question}[Which is the new final target?]
  	You remember, that we placed the result of our analysis as the final target?\newline
  	Now, we have one more final target? Which is its name?
  \end{question}
  \begin{task}
  	Place the new target in the \altverb{all} rule, use \altverb{tutorial/06_Snakefile} as the template. 
  \end{task}
  \pause
  Solution:
  \begin{lstlisting}[language=Python,style=Python]
rule all:
    input:
        "calls/all.vcf", # note the comma!
        @"plots/quals.svg"@
  \end{lstlisting}
\end{frame}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Running the final Workflow}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
	\frametitle{\HandsOn{Executing this Workflow}}
	\begin{task}
		Does our Workflow contain errors? We run a debug trial:
		\begin{lstlisting}[language=Bash, style=Shell]
$ snakemake --debug
		\end{lstlisting}
	\end{task}
	\pause
	Some targets are already present, we want the entire workflow again:
	\begin{lstlisting}[language=Bash, style=Shell]
$ snakemake -c4 --forcerun
$ # or short with
$ snakemake -c4 -F
	\end{lstlisting}
     
	\begin{question}
		What do you observe? Why \altverb{-c4}?
	\end{question}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
	\frametitle{\HandsOn{Visualising the Output}}
	On <++cluster.name++> the Linux distro is "<++cluster.distro++>", which \newline provides the \altverb{<++cluster.display_program++>}-program to display simple images. We shall invoke:
	\begin{lstlisting}[language=Bash, style=Shell]
$ <++cluster.display_program++> plots/quals.svg
	\end{lstlisting}
	The figure has no axis-labels.
	\begin{question}
		What does the figure display?
	\end{question}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
	\frametitle{Our ``final'' Workflow}
	\footnotesize{Refer to \altverb{<++course.pathtosolutions++>/06_Snakefile} as the workflow solution, so far.}
	\vspace{-1em}
	\begin{columns}
		\begin{column}{0.5\textwidth}
			\begin{lstlisting}[language=Python,style=Python,basicstyle=\tiny]
SAMPLES = ["A", "B"]
				
rule all:
	input:
		"plots/quals.svg",
		"calls/all.vcf"
				
rule bwa_map:
	input:
		"data/genome.fa",
		"data/samples/{sample}.fastq"
	output:
		"mapped_reads/{sample}.bam"
	shell:
		"bwa mem {input} | samtools view -Sb"
		" - > {output}"
				
rule samtools_sort:
	input:
	    "mapped_reads/{sample}.bam"
	output:
		"sorted_reads/{sample}.bam"
	shell:
		"samtools sort"
		" -T sorted_reads/{wildcards.sample}"
		" -O bam {input} > {output}"
			\end{lstlisting}
		\end{column}
		\begin{column}{0.5\textwidth}
			\begin{lstlisting}[language=Python,style=Python,basicstyle=\tiny]
rule samtools_index:
	input:
		"sorted_reads/{sample}.bam"
	output:
		"sorted_reads/{sample}.bam.bai"
	shell:
		"samtools index {input}"
				
rule bcftools_call:
	input:
		fa="data/genome.fa",
		bam=expand("sorted_reads/{sample}.bam",
		           sample=SAMPLES),
		bai=expand("sorted_reads/{sample}.bam.bai",
		          sample=SAMPLES)
	output:
		"calls/all.vcf"
	shell:
		"bcftools mpileup -f {input.fa}"
		" {input.bam} | "
		"bcftools call -mv - > {output}"
				
rule plot_quals:
	input:
		"calls/all.vcf"
	output:
		"plots/quals.svg"
	script:
		"scripts/plot-quals.py"
			\end{lstlisting}
		\end{column}
	\end{columns}
\end{frame}







